File: tests/runtime/smoke/host-wasmer.rs
use anyhow::Result;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Arc,
};
use wasmer::WasmerEnv;

wit_bindgen_wasmer::export!("./tests/runtime/smoke/imports.wit");

#[derive(WasmerEnv, Clone)]
pub struct MyImports {
    hit: Arc<AtomicBool>,
}

impl imports::Imports for MyImports {
    fn thunk(&mut self) {
        self.hit.store(true, Ordering::Relaxed);
        println!("in the host");
    }
}

wit_bindgen_wasmer::import!("./tests/runtime/smoke/exports.wit");

fn run(wasm: &str) -> Result<()> {
    let hit = Arc::new(AtomicBool::new(false));
    let exports = crate::instantiate(
        wasm,
        |store, import_object| {
            imports::add_to_imports(store, import_object, MyImports { hit: hit.clone() })
        },
        |store, module, import_object| exports::Exports::instantiate(store, module, import_object),
    )?;

    exports.thunk()?;

    assert!(hit.load(Ordering::Relaxed));

    Ok(())
}

File: tests/runtime/smoke/host-wasmer.py
from exports.bindings import Exports
from imports.bindings import add_imports_to_imports, Imports
from typing import Any
import sys
import wasmer # type: ignore

class MyImports:
    def thunk(self) -> None:
        self.hit = True

def run(wasm_file: str) -> None:
    store = wasmer.Store()
    module = wasmer.Module(store, open(wasm_file, 'rb').read())
    wasi_version = wasmer.wasi.get_version(module, strict=False)
    if wasi_version is None:
        import_object = {}
    else:
        wasi_env = wasmer.wasi.StateBuilder('test').finalize()
        import_object = wasi_env.generate_imports(store, wasi_version)

    wasm: Exports
    def get_export(name: str) -> Any:
        return wasm.instance.exports.__getattribute__(name)

    imports = MyImports()
    add_imports_to_imports(store, import_object, imports, get_export)
    wasm = Exports(store, import_object, module)
    wasm.thunk()
    assert(imports.hit)

if __name__ == '__main__':
    run(sys.argv[1])
